### PR TITLE
refactor(relay): split compact block execute code, more readable

### PR DESCRIPTION
**Pull request by chanhsu001**
### What problem does this PR solve?

* refactor compact block execute code, split compact block pre-check code into
non_contextual_check() and contextual_check() function
* move post process code into missing_or_collided_post_process() 
after failed to rebuild whole block, found missing or collided transactions. 
* use header reference instead of into_view() and import PendingCompactBlockMap
* add function comment in markdown format

i still not sure about function comments, please help.

Problem Summary:
the execute() function is about 250 lines, and contains some of pre-checks, and post-process after re-build block failed. 
i tried to make it more readable and hope you guys provide suggestions about how to make refactor better.

### What is changed and how it works?

What's Changed:
only split bunch of codes into 3 functions, not mean to change logic and behavior
I've run test and integration-test and passed ("Refactor Flow" should include these tests).

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None: Exclude this PR from the release note.
```

### CODE REVIEW

File: sync/src/relayer/compact_block_process.rs [hunk 1]
```diff
@@ -273,3 +156,182 @@ impl<'a> HeaderProvider for CompactBlockMedianTimeView<'a> {
...
+// check compact block is in pending
+// check compact header verification
+fn contextual_check(
+    compact_block: &CompactBlock,
+    shared: &Arc<SyncShared>,
+    active_chain: &ActiveChain,
+    nc: &Arc<dyn CKBProtocolContext>,
+    peer: PeerIndex,
+) -> Status {
+    let header = compact_block.header().into_view();
```
Comment (driftluo) [inline sync/src/relayer/compact_block_process.rs:209]: Don't always `into_view`, it's expensive(always hash itself), header must `into_view` in global, and always use reference on it

File: sync/src/relayer/compact_block_process.rs [hunk 2]
```diff
@@ -273,3 +156,182 @@ impl<'a> HeaderProvider for CompactBlockMedianTimeView<'a> {
...
+                .insert_block_status(block_hash.clone(), BlockStatus::BLOCK_INVALID);
+            return StatusCode::CompactBlockHasInvalidHeader
+                .with_context(format!("{} {}", block_hash, err));
+        }
+    }
+
+    Status::ok()
+}
+
+type PendingCompactBlockMap = HashMap<
```
Comment (driftluo) [inline sync/src/relayer/compact_block_process.rs:298]: Do not repeat type alias it, please refer to the original definition

File: sync/src/relayer/compact_block_process.rs (hunk 1 above)
Comment (chanhsu001) [reply sync/src/relayer/compact_block_process.rs]: Thank, i fixed it as your suggestion.

File: sync/src/relayer/compact_block_process.rs (hunk 2 above)
Comment (chanhsu001) [reply sync/src/relayer/compact_block_process.rs]: import it, done

File: sync/src/relayer/compact_block_process.rs [hunk 3]
```diff
@@ -273,3 +160,175 @@ impl<'a> HeaderProvider for CompactBlockMedianTimeView<'a> {
...
+        .build();
+    let message = packed::RelayMessage::new_builder().set(content).build();
+    let sending = send_message_to(nc.as_ref(), peer, &message);
+    if !sending.is_ok() {
+        ckb_logger::warn_target!(
+            crate::LOG_TARGET_RELAY,
+            "ignore the sending message error, error: {}",
+            sending
+        );
+    }
```
Comment (keroro520) [inline sync/src/relayer/compact_block_process.rs:333]: let content = packed::GetBlockTransactions::new_builder()
        .block_hash(block_hash)
        .indexes(missing_transactions.pack())
        .uncle_indexes(missing_uncles.pack())
        .build();
    let message = packed::RelayMessage::new_builder().set(content).build();
    match send_message_to(nc.as_ref(), peer, &message) {
        Ok(_) => {
            pending_compact_blocks
                .entry(block_hash.clone())
                .or_insert_with(|| (compact_block, HashMap::default(), unix_time_as_millis()))
                .1
                .insert(peer, (missing_transactions, missing_uncles));
        }
        Err(err) => {
            ckb_logger::warn_target!(
                crate::LOG_TARGET_RELAY,
                "ignore the sending message error, error: {:?}",
                err
            );
        }
    }

**Review by keroro520 — APPROVED** [review]
LGTM

**Review by driftluo — APPROVED** [review]

**Comment by chanhsu001** [comment]
bors merge=driftluo,keroro520

**Pull request merged**